### ISSUE TITLE
xm.save() stucks at last epoch

### ISSUE THREAD

**Issue by mobassir94** [pytorch/xla#2724]
i was working on this kernel [ViT - Pytorch xla (TPU) for leaf disease](https://www.kaggle.com/mobassir/vit-pytorch-xla-tpu-for-leaf-disease/) 
you can see that version 4,5 and 8 failed,the code is exactly same as version 3,i just tried to train for 10 epoch instead of 1(for single fold)

![](https://i.ibb.co/j3WkWmY/problem.png)

from the image attached above you can see that in draft / interactive session it took 1 hour 6 minutes 57 seconds to finish my model training on pytorch tpu

i just committed exact same code and even after 2 hour 8 minutes it didn't finish,stuck after 100%

in version 7 of that kernel i kept everything as it is and just stopped using xm.save() for saving weight and it worked well so it is clear that i am not able to save each and every epoch's model weight file during training, any solution please? thank you

**Comment by taylanbil** [comment]
I can look into this tomorrow more. We test `xm.save` nightly, tests should show us if it has an inherent problem. So it saves in the 1 epoch case, and fails in 10 epoch case? how many epochs does it do before hanging?

Unrelated, but took a quick peek and observed that you have unnecessary `.item()` calls inside for loop that must be slowing you down **considerably**.

example:
```python
            if running_loss is None:
                running_loss = loss.item()
            else:
                running_loss = running_loss * .99 + loss.item() * .01
```
the `item()` calls forces a premature execution of the graph and triggers a data transfer from device to cpu. I'd keep `running_loss` as a tensor, and send that to cpu once, when I absolutely have to (as in, loss reporting, or downstream calculation or something).

same comment for `loss_sum`

```python
        loss_sum += loss.item()*image_labels.shape[0]
```

No need to call `item()` there inside the for loop.

Also, I'll remind again [this recommendation](https://github.com/pytorch/xla/issues/2712#issuecomment-754182254) which will speed up the run even more.

**Comment by taylanbil** [comment]
Our tests do not suggest xm.save has a problem, fwiw. Can you provide more info, possibly a repro script?

**Comment by mobassir94** [comment]
@taylanbil thank you for your suggestions,i am applying the changes you suggested in that notebook,i have implemented another torch tpu baseline using same data,which is here : https://www.kaggle.com/mobassir/faster-torch-tpu-baseline-for-leaf-disease

this notebook is much faster but here also i am having same issue "notebook commit doesn't finish if i save weight inside epoch loop"

to reproduce you can follow these steps : 
1. fork this notebook (version 6) in kaggle : https://www.kaggle.com/mobassir/faster-torch-tpu-baseline-for-leaf-disease
2. inside train_model() function uncomment code that i have commented,
uncomment this **xm.save(model.state_dict(), os.path.join(f'{kernel_type}_fold{fold}_epoch_{epoch}_val_acc_{acc}.pth'))** or this  :
```
best_file = f'{kernel_type}_best_fold.pth'
xm.save(model.state_dict(), os.path.join(best_file))
if acc > acc_max:
    xm.save(model.state_dict(), os.path.join(best_file))
    #xser.save(model.state_dict(), os.path.join(best_file))
    acc_max = acc
```
3. now commit the notebook / click "save & run all" 

the notebook commit doesn't finish then

**Comment by mobassir94** [comment]
xm.save() is okay,there was some issue with kaggle kernels i guess, sorry for inconvenience
